resource_retriever/include/resource_retriever/retriever.h

```diff
@@ -82,6 +82,14 @@ class RESOURCE_RETRIEVER_PUBLIC Retriever
    */
   MemoryResource get(const std::string & url);
 
+  /**
+   * \brief Resolve the URL to a file on disk
+   * \param url The url to resolve to a local path on disk
+   * \return The path on disk if the url resolves to a local path, else empty string
+   * \throws resource_retriever::Exception if anything goes wrong.
```

Comment:
What about using `optional<std::string>` instead of an empty string, just so the contract is a bit more clear?

Possible edge case: I pass this function `"file://"`, and it resolves to `""`, which now intersects with the error / "ignore" mode.

Reply:
> What about using optional<std::string> instead of an empty string, just so the contract is a bit more clear?

I don't think we can use `std::optional` yet because ROS Rolling still targets the same C++ version as ROS Foxy which is C++14.

> Possible edge case: I pass this function "file://", and it resolves to "", which now intersects with the error / "ignore" mode.

If I understand correctly that case is an error, and maybe should raise an exception. It looks like `file://` is not a valid file URI according to [RFC 8089](https://tools.ietf.org/html/rfc8089). `file://` would mean what comes next is `auth-path`, and auth-path requires `path-absolute` from [RFC 3986](https://tools.ietf.org/html/rfc3986) which can't be empty; it requires a leading "/".

```
      file-URI       = file-scheme ":" file-hier-part

      file-scheme    = "file"

      file-hier-part = ( "//" auth-path )
                     / local-path

      auth-path      = [ file-auth ] path-absolute
```

and

```
      path-absolute = "/" [ segment-nz *( "/" segment ) ]
```


Also interesting is that it appears `file:` doesn't actually support relative paths. both `file:` and `file://` lead to `path-absolute` which must start with a `/`.